### ISSUE TITLE
Feature request: dom0 should be using ext4 and mounted noatime (or at least relatime)

### ISSUE THREAD

**Issue by sammcj** [xcp-ng/xcp#122]
1. It's quite scary that the dom0 is still using ext3 (I know this is hangover from XenServer) but ext3 has long since been deprecated and ext4 should be used, in fact [ext3 was removed from the standard linux kernel in 4.3][(https://www.phoronix.com/scan.php?page=news_item&px=linux-kernel-dropping-ext3)

2. Additionally ext3/4 filesystems should be mounted with noatime to reduce disk IO and improve performance, infact many modern Linux distributions have made `relatime` the default, however that is incredibly conservative as essentially no software uses atime for operations anymore, in the 'old days' atime was used by software like mail servers to determine message state.

  It has been argued that noatime should be the _default_ mount option in the Linux kernel and there is plenty of easy to find discussion on the internet around this and the performance improvements, one example: https://administratosphere.wordpress.com/2011/07/29/instant-10-20-boost-in-disk-performance-the-noatime-option/

  However, if someone can think of a good reason that XCP-ng is special and should have atime would suggest at least mounting ext3/4 filesystems with `relatime`.

  > "relative atime only updates the atime if the previous atime is older than the mtime or ctime. Like noatime, but useful for applications like mutt that need to know when a file has been read since it was last modified."

---

default `/etc/fstab`:

```
LABEL=root-mxdedt    /         ext3     defaults   1  1
LABEL=swap-mxdedt          swap      swap   defaults   0  0
LABEL=logs-mxdedt    /var/log         ext3     defaults   0  2
/opt/xensource/packages/iso/XenCenter.iso   /var/xen/xc-install   iso9660   loop,ro   0
```

proposed `/etc/fstab`:

```
LABEL=root-mxdedt    /         ext4     defaults,noatime   1  1
LABEL=swap-mxdedt          swap      swap   defaults   0  0
LABEL=logs-mxdedt    /var/log         ext4     defaults,noatime  0  2
/opt/xensource/packages/iso/XenCenter.iso   /var/xen/xc-install   iso9660   loop,ro   0
```

Note: Simply updating fstab from ext3 to ext4 will only enable ext4 features that are compatible with ext3, Obviously in addition to mounting the filesystems as ext4 although they were created as ext3 (which is supported), it would be better if they were initially created as ext4, but this would be a decent first step and perhaps convince anyone that there may be issues with ext4.

---

_This is an improvement suggestion as per [a brief discussion on the XCP-ng forums](https://xcp-ng.org/forum/post/7117) based on my XenServer and CentOS tuning experiences._

**Comment by sammcj** [comment]
Note: We've been running XenServer dom0 with noatime and mounting as ext4 for 3 or so years without issue across 32~ physical hosts and thousands of VMs.

**Comment by stormi** [comment]
So I've read https://lkml.org/lkml/2015/8/31/22 and it appears that support for ext3 has not been removed at all. Responsibility for such systems has just been transferred from the codebase of `ext3` to the codebase of `ext4` which already handled it by default in some linux distributions which compiled the kernel with `CONFIG_EXT4_USE_FOR_EXT23`. 

From what I understand:
* mount an existing `ext3` filesystem as `ext3` => same as before, though the code used if from the newer codebase
* mount an existing `ext3` filesystem as `ext4` => benefit from more features without breaking the ability to come back to `ext3` mode.

So I don't think we can say that `ext3` has been deprecated. It's still supported and part of the regression testsuites, and not going to be dropped from kernel soon.

I'm not in favour of changing the mount "driver" (technically there's only one driver now, ext4, but let's still call "ext3" a driver) from ext3 to ext4 unless there's a significant gain, but switching entirely to ext4 will probably happen sooner or later. Whether it comes from us first, or from XenServer which has much more regression-testing resources, I don't know yet. From what I understand in some comments at https://bugs.xenserver.org/browse/XSO-220, there are some issues with `ext4`. I mean, not with the filesystem itself, but with some parts of the XS stack. We need to understand them, list them, and see if that's something we can improve, or just ignore (with a warning).

**Comment by stormi** [comment]
I agree. We just need to double check and test everything before we do the switch, if XS won't do it. Would anyone know of a way to check whether atime information is being read at all, besides reading thousands of lines of code? :)

**Comment by sammcj** [comment]
Lint and she’ll-be-right ;)

(Joking, clearly)

--
Sam McLeod

**Comment by stormi** [comment]
I have written a guide to modifying the installation ISO and the installer. Maybe that can help someone experiment with a version of XCP-ng that would use EXT4.
https://github.com/xcp-ng/xcp/wiki/Modifying-the-installer

**Comment by sammcj** [comment]
I posted on the forum, probably should have posted here but mostly looking good one error I had was when trying to remove the ext4 SR:

```
# xe sr-destroy uuid=4259d179-8986-e25b-0910-ea07bff2438b
Error code: SR_BACKEND_FAILURE_202
Error parameters: , General backend error [opterr=Unsupported SR type: ext4],
```

**Comment by TylerDurden2019** [comment]
In case someone wants to test EXT4 dom0 on XCP-ng 8.0, this is what I did. 
If I did something wrong, please let me know.

**Install the EXT4 driver**
`yum install sm-additional-drivers`

**Update /etc/fstab** as per instructions provided above.
`sed -i 's/\/         ext3     defaults /\/         ext4     defaults,noatime/g' /etc/fstab'`
`sed -i 's/\/var\/log         ext3     defaults /\/var\/log         ext4     defaults,noatime/g' /etc/fstab`

`reboot` to check if all is ok.

**Enter single user mode**
`init 1`
Go to the console

**Convert /var/log**
`blkid`

Look for the device with /dev/sda5: LABEL="log-xxxxxx"

`umount /dev/sda5`
`fsck.ext3 -pf /dev/sda5`
`tune2fs -O extents,uninit_bg,dir_index /dev/sda5`
`fsck.ext4 -yfD /dev/sda5`
`mount /var/log`

**Convert the root dir /**
`blkid`

Look for the device with /dev/sda1: LABEL="root-xxxxxx"

`mount -o remount,ro /`
`fsck.ext3 -pf /dev/sda1`
`tune2fs -O extents,uninit_bg,dir_index /dev/sda1`
`fsck.ext4 -yfD /dev/sda1`
`reboot`

**Comment by De-Occultist** [comment]
In case someone wants to test ext4 dom0 on XCP-ng 8.0, this is what I did. 
If I did something wrong, please let me know.

**Install the EXT4 driver**
`yum install sm-additional-drivers`

**Update /etc/fstab** as per instructions provided above.
`sed -i 's/\/         ext3     defaults /\/         ext4     defaults,noatime/g' /etc/fstab'`
`sed -i 's/\/var\/log         ext3     defaults /\/var\/log         ext4     defaults,noatime/g' /etc/fstab`

`reboot` to check if all is ok.

**Enter single user mode**
`init 1`
Go to the console

**Convert /var/log**
`blkid`

Look for the device with /dev/sda5: LABEL="log-xxxxxx"

`umount /dev/sda5`
`fsck.ext3 -pf /dev/sda5`
`tune2fs -O extents,uninit_bg,dir_index /dev/sda5`
`fsck.ext4 -yfD /dev/sda5`
`mount /var/log`

**Convert the root dir /**
`blkid`

Look for the device with /dev/sda1: LABEL="root-xxxxxx"

`mount -o remount,ro /`
`fsck.ext3 -pf /dev/sda1`
`tune2fs -O extents,uninit_bg,dir_index /dev/sda1`
`fsck.ext4 -yfD /dev/sda1`
`reboot`
